File: backend-elixir/test/unit/hmac_and_dispatcher_test.exs
defmodule BackendElixir.Auth.HmacAndDispatcherTest do
  use ExUnit.Case, async: true

  alias BackendElixir.Auth.{Dispatcher, HmacVerifier}
  alias BackendElixir.TestSupport.TestTokenHelper

  @secret "test-secret"
  @issuer Dispatcher.local_hmac_issuer()
  @audience "backend"

  test "is_local_issuer matches hmac and rsa only" do
    assert Dispatcher.local_issuer?(Dispatcher.local_hmac_issuer())
    assert Dispatcher.local_issuer?(Dispatcher.local_rsa_issuer())
    refute Dispatcher.local_issuer?("http://localhost:8082/realms/training")
    refute Dispatcher.local_issuer?("")
  end

  test "hmac_verifier accepts valid token" do
    token = TestTokenHelper.make_hmac_token(@secret, @issuer, @audience, "42", 3600)
    verifier = HmacVerifier.new(@secret, @issuer, @audience)
    assert {:ok, claims} = HmacVerifier.verify(verifier, token)
    assert claims.sub == "42"
    assert claims.iss == @issuer
  end

  test "hmac_verifier rejects wrong secret" do
    token = TestTokenHelper.make_hmac_token(@secret, @issuer, @audience, "42", 3600)
    verifier = HmacVerifier.new("different-secret", @issuer, @audience)
    assert {:error, _reason} = HmacVerifier.verify(verifier, token)
  end

  test "hmac_verifier rejects expired token" do
    token = TestTokenHelper.make_hmac_token(@secret, @issuer, @audience, "42", -3600)
    verifier = HmacVerifier.new(@secret, @issuer, @audience)
    assert {:error, _reason} = HmacVerifier.verify(verifier, token)
  end

  test "hmac_verifier rejects wrong audience" do
    token = TestTokenHelper.make_hmac_token(@secret, @issuer, "someone-else", "42", 3600)
    verifier = HmacVerifier.new(@secret, @issuer, @audience)
    assert {:error, _reason} = HmacVerifier.verify(verifier, token)
  end

  test "hmac_verifier rejects wrong issuer" do
    token = TestTokenHelper.make_hmac_token(@secret, "unexpected-issuer", @audience, "42", 3600)
    verifier = HmacVerifier.new(@secret, @issuer, @audience)
    assert {:error, _reason} = HmacVerifier.verify(verifier, token)
  end

  test "hmac_verifier rejects unexpected algorithm" do
    token = TestTokenHelper.make_hmac_token(@secret, @issuer, @audience, "42", 3600, "HS384")
    verifier = HmacVerifier.new(@secret, @issuer, @audience)
    assert {:error, _reason} = HmacVerifier.verify(verifier, token)
  end

  test "hmac_verifier rejects alg none" do
    token = TestTokenHelper.make_alg_none_token(@issuer, @audience, "42", 3600)
    verifier = HmacVerifier.new(@secret, @issuer, @audience)
    assert {:error, _reason} = HmacVerifier.verify(verifier, token)
  end

  test "dispatcher routes by issuer and rejects unknown issuer" do
    dispatcher =
      Dispatcher.new()
      |> Dispatcher.register(@issuer, HmacVerifier, HmacVerifier.new(@secret, @issuer, @audience))

    good_token = TestTokenHelper.make_hmac_token(@secret, @issuer, @audience, "7", 3600)
    assert {:ok, claims} = Dispatcher.verify(dispatcher, good_token)
    assert claims.sub == "7"

    unknown_issuer_token = TestTokenHelper.make_hmac_token(@secret, "unknown-issuer", @audience, "7", 3600)
    assert {:error, :unknown_issuer} = Dispatcher.verify(dispatcher, unknown_issuer_token)
  end

  test "dispatcher rejects malformed token" do
    dispatcher =
      Dispatcher.new() |> Dispatcher.register(@issuer, HmacVerifier, HmacVerifier.new(@secret, @issuer, @audience))

    assert {:error, :malformed_token} = Dispatcher.verify(dispatcher, "not-a-jwt")
  end
end

File: backend-js-ts-express/tests/hmac_and_dispatcher.test.ts
// backend-js-express/tests/hmac_and_dispatcher.test.jsの型付き移植。テスト内容は変更していない。
// backend-java/HmacAndDispatcherTest.java・backend-python/test_hmac_and_dispatcher.pyの移植でもある。
// HmacVerifier単体とDispatcherの振り分けロジックを検証する(DB不要)

import test from 'node:test';
import assert from 'node:assert/strict';
import { HmacVerifier, Dispatcher, isLocalIssuer, LOCAL_HMAC_ISSUER, LOCAL_RSA_ISSUER, VerifyError } from '../src/auth/jwt';
import { makeHmacToken, makeAlgNoneToken } from './support/test_token_helper';

// HS256は最低256bit(32バイト)の鍵長が実質的に必要になるため、十分な長さのテスト用秘密鍵を使う
const SECRET = 'test-secret-at-least-32-bytes-long!!';

test('isLocalIssuer matches HMAC and RSA local issuers only', () => {
  assert.equal(isLocalIssuer(LOCAL_HMAC_ISSUER), true);
  assert.equal(isLocalIssuer(LOCAL_RSA_ISSUER), true);
  assert.equal(isLocalIssuer('http://localhost:8082/realms/training'), false);
  assert.equal(isLocalIssuer(''), false);
});

test('HmacVerifier accepts a valid token', async () => {
  const token = makeHmacToken(SECRET, LOCAL_HMAC_ISSUER, 'backend', '42', 3600);
  const verifier = new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend');
  const claims = await verifier.verify(token);
  assert.equal(claims.sub, '42');
  assert.equal(claims.iss, LOCAL_HMAC_ISSUER);
});

test('HmacVerifier rejects a token signed with the wrong secret', async () => {
  const token = makeHmacToken(SECRET, LOCAL_HMAC_ISSUER, 'backend', '42', 3600);
  const verifier = new HmacVerifier('different-secret-also-32-bytes-long!', LOCAL_HMAC_ISSUER, 'backend');
  await assert.rejects(() => verifier.verify(token), VerifyError);
});

test('HmacVerifier rejects an expired token', async () => {
  const token = makeHmacToken(SECRET, LOCAL_HMAC_ISSUER, 'backend', '42', -3600);
  const verifier = new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend');
  await assert.rejects(() => verifier.verify(token), VerifyError);
});

test('HmacVerifier rejects the wrong audience', async () => {
  const token = makeHmacToken(SECRET, LOCAL_HMAC_ISSUER, 'someone-else', '42', 3600);
  const verifier = new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend');
  await assert.rejects(() => verifier.verify(token), VerifyError);
});

test('HmacVerifier rejects the wrong issuer', async () => {
  const token = makeHmacToken(SECRET, 'some-other-issuer', 'backend', '42', 3600);
  const verifier = new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend');
  await assert.rejects(() => verifier.verify(token), VerifyError);
});

// 【セキュリティ上重要な確認】ヘッダのalgが期待(HS256)と異なる場合は、
// 仮に鍵が正しくても拒否されること(アルゴリズム混同攻撃対策)
test('HmacVerifier rejects an unexpected algorithm', async () => {
  const token = makeHmacToken(SECRET, LOCAL_HMAC_ISSUER, 'backend', '42', 3600, 'HS384');
  const verifier = new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend');
  await assert.rejects(() => verifier.verify(token), VerifyError);
});

test('HmacVerifier rejects alg=none tokens', async () => {
  const token = makeAlgNoneToken(LOCAL_HMAC_ISSUER, 'backend', '42', 3600);
  const verifier = new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend');
  await assert.rejects(() => verifier.verify(token), VerifyError);
});

test('Dispatcher routes by issuer and rejects unknown issuers', async () => {
  const dispatcher = new Dispatcher().register(LOCAL_HMAC_ISSUER, new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend'));

  const goodToken = makeHmacToken(SECRET, LOCAL_HMAC_ISSUER, 'backend', '7', 3600);
  const claims = await dispatcher.verify(goodToken);
  assert.equal(claims.sub, '7');

  const unknownIssuerToken = makeHmacToken(SECRET, 'unknown-issuer', 'backend', '7', 3600);
  await assert.rejects(() => dispatcher.verify(unknownIssuerToken), VerifyError);
});

test('Dispatcher rejects a malformed token', async () => {
  const dispatcher = new Dispatcher().register(LOCAL_HMAC_ISSUER, new HmacVerifier(SECRET, LOCAL_HMAC_ISSUER, 'backend'));
  await assert.rejects(() => dispatcher.verify('not-a-jwt'), VerifyError);
});
